### PR TITLE
[CDAP-8327] Add persist threshold to limit the number of processed metrics values in memory

### PR DESCRIPTION
**Pull request by ecfm**
https://issues.cask.co/browse/CDAP-8327

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 1]
```diff
@@ -332,9 +336,8 @@ private void processMetrics() {
         try {
           Deque<MetricValues> recordsCopy = new LinkedList<>();
           Map<TopicIdMetaKey, byte[]> topicMessageIdsCopy = new HashMap<>(topicMessageIds);
-          // TODO: (CDAP-8327) there is a risk of running out-of-memory if other threads keep writing to records
           Iterator<MetricValues> iterator = records.iterator();
-          while (iterator.hasNext()) {
+          while (iterator.hasNext() && recordsCopy.size() < persistThreshold) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: If you terminate earlier, that means the unprocessed entries will be left in the shared queue, which potentially it can grow faster than the `persistThreshold` could empty it, causing OOM

Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Also, on a related note, right now we always sleep for 1 second, which can possibly make the metrics never be able to catchup if it falls behind.

Comment (ecfm) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: I added a check before fetching new messages to ensure that no new metrics will be added to `records` when `records.size()` exceeds the threshold. Thus there will be at most 2*threshold metrics in memory.

Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Still breaking from the loop means the metrics may never be able to catchup.

Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: I think a better logic would be this:

1. Turn the shared queue into a blocking queue. The size is configurable.
2. Each thread will fetch up to (blocking queue size / number of threads) messages from TMS. Use the `put` method to enqueue to block if the queue is full.
3. The leader writer thread will only deque up to the blocking queue size (using the `poll` method).
4. The sleep should be calculated based on how much time was spent in fetching and writing (for the leader).

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 2]
```diff
@@ -288,7 +291,7 @@ private void processMetrics() {
       // If the size of records exceeds the threshold, stop fetching for new messages and adding new records.
       // Return to sleep for metricsProcessIntervalMillis to wait for metrics values to be persisted and
       // removed from records.
-      if (records.size() > persistThreshold) {
+      if (records.size() > queueSize) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Why need this check? Isn't the put call will block?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 3]
```diff
@@ -85,9 +88,12 @@
   private final MetricStore metricStore;
   private final Map<String, String> metricsContextMap;
   private final int fetcherLimit;
-  private final ConcurrentLinkedDeque<MetricValues> records;
+  private final int queueSize;
+  private final BlockingQueue<MetricValues> records;
   private final ConcurrentMap<TopicIdMetaKey, byte[]> topicMessageIds;
+  private final ConcurrentMap<TopicIdMetaKey, Long> topicLatestTime;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: You should merge this two concurrent map into one, by using a private static class that contains both the message id and time.

File: cdap-common/src/main/resources/cdap-default.xml [hunk 4]
```diff
@@ -1751,16 +1751,6 @@
   </property>
 
   <property>
-    <name>metrics.messaging.fetcher.limit</name>
```
Comment (chtyim) [inline cdap-common/src/main/resources/cdap-default.xml:1827]: This is a deprecated config. You should move this block into `cdap-default-deprecated.xml` file

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 5]
```diff
@@ -236,9 +245,9 @@ private void persistRecordsMessageIds(Deque<MetricValues> metricValues, Map<Topi
     }
   }
 
-  private void persistRecords(Deque<MetricValues> metricValues) throws Exception {
+  private void persistRecords(Queue<MetricValues> metricValues) throws Exception {
     long now = System.currentTimeMillis();
-    long lastRecordTime = metricValues.getLast().getTimestamp();
+    long lastRecordTime = Collections.max(topicLatestTime.values());
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: don't do this. You can easily track and find the largest time when inserting into the queue.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 6]
```diff
@@ -320,21 +352,34 @@ private void processMetrics() {
...
         if (!persistingFlag.compareAndSet(false, true)) {
           LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
-          return;
+          // Calculating the length of time used to emit the metrics fetched in current run.
+          // If the number of newly fetched metrics is smaller than fetcherLimit, use the sleep time after previous run.
+          // If the number of newly fetched metrics is equal to fetcherLimit, there might be more than fetcherLimit
+          // emitted during the last sleep time. Use the difference between the time of first and last metrics
+          // fetched instead.
+          long metricsEmittingTime = newMetricsCount < fetcherLimit ?
+            previousSleepMillis : topicLatestTime.get(topicIdMetaKey) - firstMetricTime;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Why use the previous sleep millis? Shouldn't the sleep time simply be determined based on the the time differences between current time and the metrics record time, as well as time spent in the processing?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 7]
```diff
@@ -345,9 +390,23 @@ private void processMetrics() {
...
           persistingFlag.set(false);
         }
+        if (recordsCopy.size() == 0) {
+          // Should never reach here;
+          return metricsProcessIntervalMillis;
+        }
+        // Calculating the length of time used to emit the metrics to be persisted.
+        long metricsEmittingTime = recordsCopy.peekLast().getTimestamp() - recordsCopy.peekFirst().getTimestamp();
+        // Estimate number of metrics emitted to be persisted per millisecond
+        double metricsEmittingSpeed = newMetricsCount / metricsEmittingTime;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Why this is needed?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java (hunk 6 above)
Comment (ecfm) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: I think the difference between current time and the metrics record time also contains the time delay between the emitter and the processor. Overestimating the time leads to underestimation of the metric emitting speed.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java (hunk 7 above)
Comment (ecfm) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: My theory is that `(process time + sleep time)*metricsEmittingSpeed = N`, where N is number of metrics to be processed in the next run. If N > fetcherLimit, unprocessed metrics will accumulate after each new run, since at most fetcherLimit metrics can be processed at each run. Therefore, we should determine the sleep time by the condition N < fetcherLimit

File: cdap-common/src/main/resources/cdap-default.xml (hunk 4 above)
Comment (ecfm) [reply cdap-common/src/main/resources/cdap-default.xml:1827]: I'm not able to find this file. Or should I create a new one?

Comment (chtyim) [reply cdap-common/src/main/resources/cdap-default.xml:1827]: https://github.com/caskdata/cdap/blob/develop/cdap-docs/tools/cdap-default/cdap-default-deprecated.xml

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 8]
```diff
@@ -207,8 +218,10 @@ protected void run() {
         Thread.currentThread().interrupt();
       }
     }
+    long lastRecordTime = TimeUnit.MILLISECONDS.toSeconds(records.peekLast().getTimestamp());
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: will the `records` queue empty?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 9]
```diff
@@ -236,25 +250,23 @@ private void persistRecordsMessageIds(Deque<MetricValues> metricValues, Map<Topi
     }
   }
 
-  private void persistRecords(Deque<MetricValues> metricValues) throws Exception {
-    long now = System.currentTimeMillis();
-    long lastRecordTime = metricValues.getLast().getTimestamp();
-    long delay = now - TimeUnit.SECONDS.toMillis(lastRecordTime);
+  private void persistRecords(Queue<MetricValues> metricValues, long lastRecordTime, long delay) throws Exception {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Why the delay is pass in?? Seems like this is an unrelated change?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 10]
```diff
@@ -320,25 +343,37 @@ private void processMetrics() {
...
-          // TODO: (CDAP-8327) there is a risk of running out-of-memory if other threads keep writing to records
           Iterator<MetricValues> iterator = records.iterator();
-          while (iterator.hasNext()) {
+          while (iterator.hasNext() && recordsCopy.size() < queueSize) {
             recordsCopy.add(iterator.next());
             iterator.remove();
           }
-          persistRecordsMessageIds(recordsCopy, topicMessageIdsCopy);
+          long lastRecordTime = TimeUnit.MILLISECONDS.toSeconds(recordsCopy.peekLast().getTimestamp());
+          long delay = System.currentTimeMillis() - lastRecordTime;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: the delay should only depends on record seen by this thread. In fact, the logic should be quite simple as:

```java
long startTime = System.currentMillis();
// Fetch messages from TMS and put it to the shared queue
// During the looping over the fetched messages, also get the lastMetricTimeMillis
...
// If nothing was fetched, just return
if (records.isEmpty()) {
  return metricProcessIntervalMillis;
}

if (persistingFlag.compareAndSet(false, true)) {
  persistRecords(...);   // this is a new private method for the writer to persist data
}

long endTime = System.currentMillis();
if (endTime - lastMetricTimeMillis > maxDelayMillis) {
  // Don't sleep if falling behind
  return 0L;
} else {
  long timeSpent = endTime - startTime;
  return Math.max(0L, metricsProcessIntervalMillis - timeSpent);
}
```

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 11]
```diff
@@ -222,12 +230,26 @@ protected void triggerShutdown() {
...
   private void persistRecordsMessageIds(Deque<MetricValues> metricValues, Map<TopicIdMetaKey, byte[]> messageIds) {
+    if (metricValues.isEmpty()) {
+      return;
+    }
     try {
-      if (!metricValues.isEmpty()) {
-        persistRecords(metricValues);
-      }
+      long now = System.currentTimeMillis();
+      long lastRecordTime = TimeUnit.MILLISECONDS.toSeconds(metricValues.peekLast().getTimestamp());
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: The timestamp in `MetricValue` is already in seconds.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 12]
```diff
@@ -222,12 +230,26 @@ protected void triggerShutdown() {
...
+      long delay = now - lastRecordTime;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: This is wrong. `now` is in milliseconds, while `lastRecordTime` is in seconds.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 13]
```diff
@@ -303,7 +321,8 @@ private void processMetrics() {
             try {
               payloadInput.reset(input.getPayload());
               MetricValues metricValues = recordReader.read(decoder, recordSchema);
-              records.add(metricValues);
+              records.put(metricValues);
+              lastMetricTimeMillis = TimeUnit.MILLISECONDS.toSeconds(metricValues.getTimestamp());
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: This conversion is wrong. You might want to do `TimeUnit.SECONDS.toMillis`?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 14]
```diff
@@ -318,36 +337,58 @@ private void processMetrics() {
...
-            iterator.remove();
-          }
-          persistRecordsMessageIds(recordsCopy, topicMessageIdsCopy);
-        } catch (Exception e) {
-          LOG.error("Failed to persist consumed messages.", e);
-        } finally {
-          // Set persistingFlag back to false after persisting completes.
-          persistingFlag.set(false);
+        long endTime = System.currentTimeMillis();
+        if (endTime - lastMetricTimeMillis > maxDelayMillis) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: The `lastMetricTimeMillis` was not initialized. This mean if there is no messages being fetched, then `lastMetricTimeMillis` would be some old timestamp (potentially 0) and this is always true, resulting in polling TMS without sleep?

Comment (ecfm) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Added a check to see whether new metrics are fetched

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 15]
```diff
@@ -222,12 +230,26 @@ protected void triggerShutdown() {
...
+                         ImmutableList.of(
+                           new MetricValue("metrics.process.count", MetricType.COUNTER, metricValues.size()),
+                           new MetricValue("metrics.process.delay.ms", MetricType.GAUGE, delay))));
+      metricStore.add(metricValues);
+      recordsProcessed += metricValues.size();
+      PROGRESS_LOG.debug("{} metrics records persisted. Last metric record's timestamp: {}. " +
+                           "Metrics process delay: {}ms", recordsProcessed, lastRecordTime, delay);
       try {
-        metaTable.saveMessageIds(messageIds);
+        if (!messageIds.isEmpty()) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: When will this be empty?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 16]
```diff
@@ -316,38 +335,62 @@ private void processMetrics() {
 
         if (currentMessageId != null) {
           topicMessageIds.put(topicIdMetaKey, currentMessageId);
+        } else {
+          return metricsProcessIntervalMillis; // sleep for the max amount of time allowed since no message is fetched
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: I think you still need to call `tryPersist` even though there is no message fetched for this partition. Otherwise, you may be delaying persisting metrics. E.g. considering you have two threads, the following might happen:

1. Thread 1 fetched some metrics
2. Thread 1 call `tryPersist` and successfully acquired the `persistingFlag`.
3. Thread 1 copied the records
4. Thread 2 fetched some metrics
5. Thread 2 call `tryPersist`. Since thread is still holding the flag, it will just return
6. Thread 1 persisted the metrics and release the flag.
7. Both Thread 1 and Thread 2 have empty fetch.

You can see that metrics fetched by thread 2 in step 4 will not get persisted with your updated logic.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java (hunk 15 above)
Comment (ecfm) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: This can happen if exception is thrown but not caught when adding metrics to `records` while `messageIds` is still empty, and `persistRecordsMessageIds` is called at the end of the `run()` method

Comment (ecfm) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: It's also possible when the current thread fetches nothing, but the other thread keeps adding metrics into `records`

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 17]
```diff
@@ -302,8 +329,9 @@ private void processMetrics() {
             RawMessage input = iterator.next();
             try {
               payloadInput.reset(input.getPayload());
-              MetricValues metricValues = recordReader.read(decoder, recordSchema);
-              records.add(metricValues);
+              MetricValues metricValues = metricReader.read(decoder, metricSchema);
+              metricsFromAllTopics.put(metricValues);
+              lastMetricTimeMillis = TimeUnit.SECONDS.toMillis(metricValues.getTimestamp());
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: You can instead keep the `lastMetricTimeSecs` to avoid conversion inside the loop.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 18]
```diff
@@ -318,35 +346,57 @@ private void processMetrics() {
...
         }
 
-        // Skip persisting if the records is empty
-        if (records.isEmpty()) {
-          return;
+        // Try to persist metrics and messageId's of the last metrics to be persisted if no other thread is persisting
+        tryPersist();
+
+        long endTime = System.currentTimeMillis();
+        if (endTime - lastMetricTimeMillis > maxDelayMillis) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: So if nothing was fetched, the `lastMetricTimeMillis` would still be some older timestamp, hence returning `0L` and causing a non-sleep polling. Is that the intention??

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 19]
```diff
@@ -318,35 +346,57 @@ private void processMetrics() {
...
-          LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
-          return;
+    /**
+     * Persist metrics and messageId's of the last metrics to be persisted if no other thread is persisting
+     */
+    private void tryPersist() {
+      // Ensure there's only one thread can persist metricsFromAllTopics and messageId's.
+      // If persistingFlag is false, set it to true and start persisting. Otherwise, log and return.
+      if (!persistingFlag.compareAndSet(false, true)) {
+        LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: This message is a bit misleading. It is not cannot persist. It is this thread is not the persisting thread, hence letting other threads to do the persisting of metrics.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 20]
```diff
@@ -318,35 +346,57 @@ private void processMetrics() {
...
+        return;
+      }
+      try {
+        // Skip persisting if metricsFromAllTopics is empty
+        if (metricsFromAllTopics.isEmpty()) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Don't check the concurrent queue. You should always copy to a local list first. If the copied list is empty, then you don't need to persist. Just keep the old logic with a check before calling `persistMetricsMessageIds`

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 21]
```diff
@@ -207,8 +215,10 @@ protected void run() {
         Thread.currentThread().interrupt();
       }
     }
-    // Persist records and messageId's after all ProcessMetricsThread's complete.
-    persistRecordsMessageIds(records, topicMessageIds);
+    // Persist metricsFromAllTopics and messageId's after all ProcessMetricsThread's complete.
+    if (!metricsFromAllTopics.isEmpty()) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: I think I made a comment about don't check on the concurrent queue.

Comment (ecfm) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Here all threads already complete and no thread is writing to the queue. Is it still necessary to make a copy?

Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: I see. That's ok.

**Comment by chtyim** [comment]
LGTM. Please squash before merging.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 22]
```diff
@@ -221,40 +235,61 @@ protected void triggerShutdown() {
...
-
-    PROGRESS_LOG.debug("{} metrics records processed. Last metric record's timestamp: {}. Metrics process delay: {}",
-                       recordsProcessed, lastRecordTime, delay);
+  /**
+   * Persist messageId's of the last persisted metrics of each topic into metrics meta table
+   *
+   * @param messageIds   a map with each key {@link TopicIdMetaKey} representing a topic and messageId's
+   *                     of the last persisted metric of the topic
+   */
+  private void persistLastMessageIds(Map<TopicIdMetaKey, byte[]> messageIds) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Just call it `persistMessageIds`, last of not doesn't matter.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 23]
```diff
@@ -221,40 +235,61 @@ protected void triggerShutdown() {
...
+    try {
+      // messageIds can be empty if the current thread fetches nothing while other threads keep fetching new metrics
+      // and haven't updated messageId's of the corresponding topics
+      if (!messageIds.isEmpty()) {
+        metaTable.saveMessageIds(messageIds);
+        for (Map.Entry<TopicIdMetaKey, byte[]> entry : messageIds.entrySet()) {
+          LOG.info("Persist {} in topic {}", Bytes.toStringBinary(entry.getValue()),
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Log this as debug or trace, otherwise you could be logging a lot.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 24]
```diff
@@ -221,40 +235,61 @@ protected void triggerShutdown() {
...
+                   entry.getKey().getTopicId().getTopic());
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to persist messageId's of consumed messages.", e);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Since this is non-fatal, log this as WARN.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 25]
```diff
@@ -318,35 +366,57 @@ private void processMetrics() {
...
-        } finally {
-          // Set persistingFlag back to false after persisting completes.
-          persistingFlag.set(false);
+        // Persist the copy of metrics and MessageId's only if the copy of metrics is not empty
+        if (!metricsCopy.isEmpty()) {
+          persistMetricsMessageIds(metricsCopy, topicMessageIdsCopy);
         }
       } catch (Exception e) {
-        LOG.warn("Failed to process metrics. Will be retried in next iteration.", e);
+        LOG.error("Failed to persist consumed messages.", e);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Why the change? Is it still going to be retried in next iteration?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 26]
```diff
@@ -207,8 +215,14 @@ protected void run() {
...
-    // Persist records and messageId's after all ProcessMetricsThread's complete.
-    persistRecordsMessageIds(records, topicMessageIds);
+    // Persist metricsFromAllTopics and messageId's after all ProcessMetricsThread's complete.
+    // No need to make a copy of metricsFromAllTopics and topicMessageIds because no thread is writing to them
+    if (!metricsFromAllTopics.isEmpty()) {
+      persistMetricsMessageIds(metricsFromAllTopics, topicMessageIds);
+    }
+    // Persist topicMessageIds even when metricsFromAllTopics is empty, because topicMessageIds can be updated after
+    // metricsFromAllTopics is persisted
+    persistLastMessageIds(topicMessageIds);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: Better just use the old logic. It's cleaner (check if the provided queue is empty inside the `persistMetricsMessageIds` method). Remember, less code and conditions are better.

Comment (ecfm) [reply cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java]: we still need `persistLastMessageIds(topicMessageIds);` here even when the queue is empty and persisting messageId's is skipped in `persistMetricsMessageIds`

**Comment by chtyim** [comment]
LGTM
